refactor: change mine tab to custom tab

janbritz opened the pull request:
Closes #106.

Wie besprochen, gibt es nun statt der `Mine`-Kategorie die `Custom`-Kategorie (s. #106). In dieser Kategorie werden eigene Pakete angezeigt und andere lokale Pakete, wenn in der `qtype_questionpy_visibility` ein Eintrag mit dem aktuellen Kontext zu finden ist. Später soll über die UI die Sichtbarkeit der eigenen Pakete geändert werden können.

Zusätzlich zu `ismine` wird nun auch das Feld `isfromserver` für jede Paketversion zurückgegeben. Dieses Feld ist `true`, wenn die Paketversion auch vom Application-Server angeboten wird und kann genutzt werden, um lokale Paketversionen gesondert zu markieren.

Zuvor wurden Paketversionen, welche von mehreren Quellen (Usern/Application Server) hochgeladen wurden, mehrmals zurückgegeben - das ist nun gefixed.

janbritz committed:
refactor: search w/o context and custom tab

larsbonczek commented on classes/external/favourite_package.php:
```diff
@@ -129,10 +93,10 @@ public static function execute(int $packageid, bool $favourite, int $contextid):
 
         // Mark package as favourite if it is accessible by the user.
-        if ($accessible = self::package_is_accessible($params['packageid'], $context)) {
+        if ($exists = $DB->record_exists('qtype_questionpy_pkgversion', ['packageid' => $params['packageid']])) {
```
Muss nicht eigentlich überprüft werden, ob der User das Paket in irgendeinem Kontext sehen kann?

larsbonczek commented on classes/package/package_version.php:
```diff
@@ -72,48 +109,87 @@ public static function get_by_id(int $pkgversionid): package_version {
         global $DB;
         $packages = [];
-        $records = $DB->get_records('qtype_questionpy_pkgversion', $conditions);
+        [$joinsql, $joinparams] = self::sql_get('s2.id IS NOT NULL');
```
Hier den Alias s2 so als magic value zu verwenden finde ich riskant. In jedem Fall wäre es gut, wenn in dem docstring von sql_get dokumentiert ist, welche Felder/Aliase verfügbar sind. Vor allem, da die Funktion public ist.

larsbonczek commented on classes/package/package_version.php:
```diff
@@ -72,48 +109,87 @@ public static function get_by_id(int $pkgversionid): package_version {
+        // Delete a source of the package version.
+        if ($asuser) {
+            if (!$this->ismine) {
+                throw new moodle_exception('');
```
Bitte hier einen gültigen errorcode angeben (für den ein string existiert). Es gibt zum Beispiel "nopermissions". Aber ein eigener string wäre besser.

larsbonczek commented on classes/package/package_version.php:
```diff
@@ -50,6 +50,42 @@ class package_version {
+     * @param array $params
+     * @return array A list containing the constructed sql fragment and an array of parameters.
+     */
+    public static function sql_get(string $where = '', array $params = []): array {
```
Im Grunde könntest du hier auch direkt den Query ausführen und die Ergebnisse casten, oder? Ich würde dann vorschlagen, dass du eine Funktion `get_many_where: array` und eine Funktion `get_where: ?package_version`  anbietest, die du dann in den anderen Funktionen aufrufst.

larsbonczek commented on classes/external/search_packages.php:
```diff
@@ -459,25 +405,25 @@ public static function execute(string $query, array $tags, string $category, str
+        if (is_null($params['contextid'])) {
+            self::validate_context(context_system::instance());
+        } else {
+            self::validate_context(context::instance_by_id($params['contextid'], IGNORE_MISSING));
```
Wieso IGNORE_MISSING? Macht hier keinen Unterschied, weil validate_context auch nen Fehler wirft wenn der Context leer ist. Aber dann kannst du es auch weglassen.

janbritz replied:
[Hier](https://github.com/questionpy-org/moodle-qtype_questionpy/pull/74#discussion_r1416950428) ist meine damalige Begründung.

larsbonczek commented on classes/external/remove_packages.php:
```diff
@@ -54,9 +54,9 @@ public static function execute(): array {
         $transaction = $DB->start_delegated_transaction();
```
Hier fehlt noch ein validate_context und eine Überprüfung der Berechtigung, oder nicht??

janbritz replied:
Ja, allerdings sollen die Services nicht Teil des Production-Codes werden.

larsbonczek commented on classes/external/load_packages.php:
```diff
@@ -63,16 +63,16 @@ public static function execute(): array {
         $transaction = $DB->start_delegated_transaction();
```
Hier fehlt noch ein validate_context und eine Überprüfung der Berechtigung, oder nicht??

janbritz replied:
https://github.com/questionpy-org/moodle-qtype_questionpy/pull/107#discussion_r1539091287

larsbonczek commented on db/install.xml:
```diff
@@ -54,19 +54,29 @@
+        <FIELD NAME="id" TYPE="int" LENGTH="10" SEQUENCE="true" NOTNULL="true"/>
+        <FIELD NAME="pkgversionid" TYPE="int" LENGTH="10" SEQUENCE="false" NOTNULL="true"/>
         <FIELD NAME="timecreated" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="false"/>
         <FIELD NAME="userid" TYPE="int" LENGTH="10" NOTNULL="false" SEQUENCE="false"/>
```
Ich finde es ein bisschen unsauber hier das userid Feld einfach auf null zu setzen um zu signalisieren, dass die Paketversion vom Server kommt. Mein Vorschlag wäre, das als Boolean-Flag zu speichern entweder in einem qtype_questionpy_source Datensatz oder direkt in der Paketversion.

larsbonczek reviewed (changes requested):
Ich hatte es im letzten Meeting so verstanden, dass unter "custom" alle Paketversionen angezeigt werden sollen, die von mir oder von anderen Nutzer*innen _in einem Kontext, auf den ich Zugriff habe_ hochgeladen wurden. Jetzt gibst du ja einfach alle nutzerdefinierten Pakete zurück, oder?

Du könntest bei der Gelegenheit auch mal die Suchfunktion in eine eigene Klasse refaktorisieren, damit man sie theoretisch auch aus dem PHP-Code heraus benutzen kann und nicht nur als external service.

janbritz commented:
Antwort auf [deinen Kommentar](https://github.com/questionpy-org/moodle-qtype_questionpy/pull/107#pullrequestreview-1958213249) @larsbonczek. (Leider kann man ja nicht direkt antworten.)
> [...] unter "custom" alle Paketversionen angezeigt werden sollen, die von mir oder von anderen Nutzer*innen _in einem Kontext, auf den ich Zugriff habe_ hochgeladen wurden. [...]

Ich habe es so verstanden, dass alle lokalen Pakete unter "Custom" zu finden sein sollen, da mit dem zitierten Ansatz das Problem, welches ich beim letzten Meeting angesprochen habe, bestehen bleibt:

Person X hat Zugriff auf die Kurse A und B und lädt im Kurs A ein Paket hoch.
Person Y hat auch Zugriff auf die Kurse A und B.
Person Z hat nur Zugriff auf den Kurs B und kann damit, im Gegensatz zu Person X und Y, das Paket nicht finden - Person X müsste das Paket also auch im Kurs B wieder hochladen und das ist nicht gewollt.

janbritz committed:
refactor: use context und introduce visibility

A local package can be found if there exists an entry in the visibility
table of the package with the current context of the user or if the user
uploaded that package.

janbritz committed:
fix: run tests in separate processes

and add missing docstring

janbritz committed:
fix: delegated transaction
